### PR TITLE
`<deque>`: Make `deque` support incomplete element types

### LINKED ISSUE #108: <deque>: deque should support incomplete value types
**Describe the bug**
Due to these two lines, deque cannot be used without the definition of the element type:
https://github.com/microsoft/STL/blob/ee1ede2c6207fb2be2c870bcddf67c48d9d32a24/stl/inc/deque#L250
https://github.com/microsoft/STL/blob/ee1ede2c6207fb2be2c870bcddf67c48d9d32a24/stl/inc/deque#L593

Seems unnecessary, since the enumerator doesn't seem to be referenced anywhere in the codebase.

**Command-line test case**
STL version (git commit or Visual Studio version):
ee1ede2c6207fb2be2c870bcddf67c48d9d32a24

```
#include <deque>

class foo
{
    foo();
    foo(const foo&) = delete;
    foo(foo&&) = delete;
    foo& operator=(const foo&) = delete;
    foo& operator=(foo&&) = delete;

    ~foo();

    struct impl;
    std::deque<impl> m_impl;
};

```

```
c:\Users\Davide\Desktop>cl repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.21.27702.2 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.21.27702\include\deque(606): error C2027: use of undefined type 'foo::impl'
repro.cpp(13): note: see declaration of 'foo::impl'
repro.cpp(14): note: see reference to class template instantiation 'std::deque<foo::impl,std::allocator<_Ty>>' being compiled
        with
        [
            _Ty=foo::impl
        ]
```

**Expected behavior**
I would expect it to be pimpl-friendly just like other stl implementations: https://godbolt.org/z/JNqc_R

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #108.

* Replaces every occurrence of `_EEN_DS` in STL.natvis with a detailed expression that calculates the block size from `sizeof(value_type)`.
* Every `_Block_size` is turned into a local variable, so that it would not be instantiated eagerly.
* No non-`_Ugly` identifier is introduced.
* I think no ABI break is introduced, but I'm not very sure.
  * #1307 added static data members and it seems OK to change them.

### CODE REVIEW

File: stl/inc/deque [hunk 1]
```diff
@@ -554,20 +551,19 @@ public:
...
-private:
-    static constexpr size_t _Bytes = sizeof(value_type);
-
-public:
-    static constexpr int _Block_size = _Bytes <= 1 ? 16
-                                     : _Bytes <= 2 ? 8
-                                     : _Bytes <= 4 ? 4
-                                     : _Bytes <= 8 ? 2
-                                                   : 1; // elements per block (a power of 2)
+    static constexpr int _Get_block_size() noexcept { // elements per block (a power of 2)
```
Comment (miscco) [inline stl/inc/deque:554]: This should be `_NODISCARD` and possibly `consteval`

File: tests/std/tests/GH_000108_deque_incomplete_value_types/test.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <assert.h>
+#include <deque>
+
+using namespace std;
+
+void test_gh_000108();
```
Comment (miscco) [inline tests/std/tests/GH_000108_deque_incomplete_value_types/test.cpp:9]: Can we put each test into a respective function with a proper name that explains what it is doing?

File: stl/inc/deque [hunk 3]
```diff
@@ -43,8 +41,9 @@ public:
         : _Mycont(static_cast<const _Mydeque*>(_Pdeque)), _Myoff(_Off) {}
 
     _NODISCARD reference operator*() const noexcept {
-        _Size_type _Block = _Mycont->_Getblock(_Myoff);
-        _Size_type _Off   = _Myoff % _Block_size;
+        constexpr int _Block_size = _Mydeque::_Get_block_size();
+        _Size_type _Block         = _Mycont->_Getblock(_Myoff);
```
Comment (miscco) [inline stl/inc/deque:45]: While you are there both could be `const`

File: tests/std/tests/GH_000108_deque_incomplete_value_types/test.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <assert.h>
+#include <deque>
+
+using namespace std;
+
+void test_gh_000108_push_pop_size();
```
Comment (miscco) [inline tests/std/tests/GH_000108_deque_incomplete_value_types/test.cpp:9]: Is there a reason for the separate declaration and definition of that function? We conventionally only define it before the main function

File: stl/inc/deque (hunk 3 above)
Comment (frederick-vs-ja) [reply stl/inc/deque]: I made index variables `const` when possible. But I'm not sure if there are some other variables expected to be made `const`.

File: tests/std/tests/GH_000108_deque_incomplete_value_types/test.cpp (hunk 4 above)
Comment (frederick-vs-ja) [reply tests/std/tests/GH_000108_deque_incomplete_value_types/test.cpp]: I partially duplicated the test.cpp file from that of Dev10_860421.

To be honest I'm not familiar with the tests of MSVC STL and don't know what are expected to be tested...

**Comment by StephanTLavavej** [comment]
Thanks for looking into this, but I don't believe that we should merge this PR, even if there's a way to make the visualizers work. As mentioned in the README's [Non-Goals](https://github.com/microsoft/STL#non-goals) section, we generally avoid adding non-Standard extensions, especially significant ones. The rationale is that such extensions have maintenance costs, especially as the Standard continues to evolve, and they don't help users write portable code (in fact, they can lead to users writing code that works for MSVC, only to discover that it's non-portable - there are already many ways to do that, we just try to avoid adding more).

**Comment by FrankHB** [comment]
> we generally avoid adding non-Standard extensions, especially significant ones

This is an QoI issue at current. Given the usefulness of the feature and the feasibility in fresh implementations, it may be proposed in some future version of the standard, so it may be potentially beneficial.

(And if it is not, to make the code portable, users are enforced to do the work similar in this patch to replace the standard one.)

>  such extensions have maintenance costs, especially as the Standard continues to evolve

The changes in the patch apparently incur some costs, but the effects on the cost of long term maintenance is not definite, "especially as the Standard continues to evolve".

> they don't help users write portable code

They don't help at current, until the guarantees are mandated in the future, which is at least, possible.

> in fact, they can lead to users writing code that works for MSVC, only to discover that it's non-portable - there are already many ways to do that, we just try to avoid adding more

This is probably the easiest part here. The implementation can just add checks to enforce the completeness of the types by default. Then, optionally, provide an configurable macro as an extension to disable the checks, with documented guarantees and caveats on the risks of the portability.

**Comment by frederick-vs-ja** [comment]
@StephanTLavavej  Recently I found that libc++ has an ABI-incompatible mode (`_LIBCXX_ABI_VERSION == 2`) where `deque` is made to support incomplete element types. Once clang switches to this mode by default, it becomes practical to standardize such support.

**Comment by CaseyCarter** [comment]
> Thanks for looking into this, but I don't believe that we should merge this PR, even if there's a way to make the visualizers work. As mentioned in the README's [Non-Goals](https://github.com/microsoft/STL#non-goals) section, we generally avoid adding non-Standard extensions, especially significant ones. The rationale is that such extensions have maintenance costs, especially as the Standard continues to evolve, and they don't help users write portable code (in fact, they can lead to users writing code that works for MSVC, only to discover that it's non-portable - there are already many ways to do that, we just try to avoid adding more).

Beyond our normal concern about extension behaviors, I don't want to add any more support for incomplete element types in Standard containers until WG21 solves the problem of constraining Special Member Functions for such a container. Today, `vector<list<unique_ptr>>` (for example) is unusable. `list` has a potentially-throwing move constructor, and lies about being copyable (this is what "unconstrained special member functions" means), so `vector` chooses to (try to) copy `list<unique_ptr>`s on reallocate instead of move. Since `unique_ptr` - and therefore `list<unique_ptr>` is in fact _not_ copyable we get a compile error.

We can't constrain the special member functions without unavoidable breakage for the so-called "cyclic" incomplete element type case:
```c++
struct S {
  std::container<S> some_structs;
};
```
If container attempts to determine if `S` is copyable, the compiler must first determine if `container<S>` is copyable, which requires knowing if `S` is copyable, ad infinitum.

**Comment by StephanTLavavej** [comment]
We talked about this at the weekly maintainer meeting and have consensus to close this PR. If the Standard is changed in the future to require `deque<Incomplete>` then merging such a change will be possible.

**Pull request closed without merging**